base/iobuffer.jl

```diff
@@ -489,6 +489,76 @@ function take!(io::IOBuffer)
     return data
 end
 
+"Internal method. This method can be faster than takestring!, because it does not
+reset the buffer to a usable state, and it does not check for io.reinit.
+Using the buffer after calling unsafe_takestring! may cause undefined behaviour.
+This function is meant to be used when the buffer is only used as a temporary
+string builder, which is discarded after the string is built."
+function unsafe_takestring!(io::IOBuffer)
+    start = io.seekable ? io.offset + 1 : io.ptr
+    nbytes = io.size - start + 1
+    iszero(nbytes) && return ""
+    # The C function can only copy from the start of the memory.
+    # Fortunately, in most cases, the offset will be zero.
+    return if isone(start)
+        ccall(:jl_genericmemory_to_string, Ref{String}, (Any, Int), io.data, nbytes)
+    else
+        mem = StringMemory(nbytes)
+        unsafe_copyto!(mem, 1, io.data, start, nbytes)
+        unsafe_takestring(mem)
+    end
+end
+
+"""
+    takestring!(io::IOBuffer) -> String
+
+Return the content of `io` as a `String`, resetting the buffer to its initial
+state.
+This is preferred over calling `String(take!(io))` to create a string from
+an `IOBuffer`.
+
+# Examples
+```jldoctest
+julia> io = IOBuffer();
+
+julia> write(io, [0x61, 0x62, 0x63]);
+
+julia> s = takestring!(io)
+"abc"
+
+julia> isempty(take!(io)) # io is now empty
+true
+```
+
+!!! compat "Julia 1.12"
+    This function requires at least Julia 1.12.
+"""
+function takestring!(io::IOBuffer)
+    # If the buffer has been used up and needs to be replaced, there are no bytes, and
+    # we can return an empty string without interacting with the buffer at all.
+    io.reinit && return ""
+
+    s = unsafe_takestring!(io)
```

Comment:
If the buffer is not writable, this must copy the data into a `StringMemory`, see line 471 in the `take!` function.

Reply:
I don't think so, but perhaps I'm wrong. The reason it copies in line 471 is because, if the buffer is not writable, `io.reinit` is not set to `true`, and thus the returned value is an alias of the memory in the buffer. However, this is only safe is one can guarantee the returned `Vector` is never mutated. We can guarantee that when returning `String`, so it's fine that the returned string shares memory with the unwriteable buffer.

Reply:
This will normally be fine, but the following example seg faults with this PR.

```julia
iow = IOBuffer()
write(iow, zeros(UInt8, 10_000_000))
v1 = take!(iow)
ior = IOBuffer(v1)
s1 = takestring!(ior)
s1 = nothing
GC.gc()
GC.gc()
GC.gc()
v1[1]
```

This is because after a call to `unsafe_takestring(m::Memory{UInt8})` `m` becomes unsafe to read because its memory may get freed by Julia if the string isn't needed.